src/main/java/fi/aalto/cs/apluscourses/model/ModuleInstallerImpl.java

```diff
@@ -0,0 +1,133 @@
+package fi.aalto.cs.apluscourses.model;
+
+import fi.aalto.cs.apluscourses.utils.async.TaskManager;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.jetbrains.annotations.NotNull;
+
+public class ModuleInstallerImpl<T> implements ModuleInstaller {
+
+  private final ModuleSource moduleSource;
+  private final TaskManager<T> taskManager;
+
+  public ModuleInstallerImpl(ModuleSource moduleSource, TaskManager<T> taskManager) {
+    this.moduleSource = moduleSource;
+    this.taskManager = taskManager;
+  }
+
+  protected T waitUntilLoadedAsync(Module module) {
+    return taskManager.fork(() -> module.stateMonitor.waitUntil(Module.LOADED));
+  }
+
+  protected T installInternalAsync(Module module) {
+    ModuleInstallation moduleInstallation = new ModuleInstallation(module);
+    return taskManager.fork(moduleInstallation::doIt);
+  }
+
+  /**
+   * Installs multiple modules and their dependencies.
+   *
+   * @param modules A {@link List} of {@link Module}s to be installed.
+   */
+  public void install(@NotNull List<Module> modules) {
+    taskManager.joinAll(modules
+        .stream()
+        .map(this::installInternalAsync)
+        .collect(Collectors.toList()));
+  }
+
+  /** Installs a module and its dependencies.
+   *
+   * @param module A {@link Module} to be installed.
+   */
+  public void install(Module module) {
+    taskManager.join(installInternalAsync(module));
+  }
+
+  @Override
+  public void installAsync(@NotNull List<Module> modules) {
+    taskManager.fork(() -> install(modules));
+  }
+
+  private class ModuleInstallation {
+
+    private final Module module;
+    List<Module> dependencies;
+
+    public ModuleInstallation(Module module) {
+      this.module = module;
+    }
+    
+    public void doIt() {
+      try {
+        fetch();
+        load();
+        waitForDependencies();
+      } catch (IOException | ModuleLoadException e) {
+        module.stateMonitor.set(Module.ERROR);
```

Comment:
are you sure we do not need those `Exceptions` themselves somehow?

Reply:
Yep, you are right, we should show the user what went wrong and allow them to clean the modules state somehow and try again. But I'd suggest we implement that later (in particular after/during implementing #64 when we have more complete understanding on mapping between IDEA modules and `Module` objects and have designed policies like "what happens if user removes or unloads imported module", "what if user creates manually a module called GoodStuff and then try to import that from the grader").